### ISSUE TITLE
Confused about "return  " at plugin/futr3d/models/backbones/radar_encoder.py#L93

### ISSUE THREAD

**Issue by LeopoldACC** [Tsinghua-MARS-Lab/futr3d#12]
Hi， @xyaochen ，thanks for your work in multi sensor fusion detection.
But I'm confused about the implementation in [plugin/futr3d/models/backbones/radar_encoder.py#L93](https://github.com/Tsinghua-MARS-Lab/futr3d/blob/5c290b699ddc5098141414921da6d0d3fbad4b5e/plugin/futr3d/models/backbones/radar_encoder.py#L93)
If implement like that, the module registried as "radar_encoder" will return nothing.Can you help me understand it?Thanks a lot!
```python
    def forward(self, points):
        '''
        points: [B, N, C]. N: as max
        masks: [B, N, 1]
        ret: 
            out: [B, N, C+1], last channel as 0-1 mask
        '''
        masks = points[:, :, [-1]]
        x = points[:, :, :-1]
        xy = points[:, :, :2]
        
        for feat_layer in self.feat_layers:
            x = feat_layer(x)
        
        out = x * masks

        out = torch.cat((x, masks), dim=-1)

        out = torch.cat((xy, out), dim=-1)
        return 
```

**Comment by xyaochen** [comment]
Sorry for the typo! It should be "return out"

**Comment by LeopoldACC** [comment]
# code's change
```bash
diff --git a/plugin/futr3d/configs/cam_radar/res101_radar.py b/plugin/futr3d/configs/cam_radar/res101_radar.py
index 79bfd48..6adab02 100644
--- a/plugin/futr3d/configs/cam_radar/res101_radar.py
+++ b/plugin/futr3d/configs/cam_radar/res101_radar.py
@@ -1,6 +1,10 @@
+# _base_ = [^M
+#     '../../../../configs/_base_/datasets/nus-3d.py',
+#     '../../../../configs/_base_/default_runtime.py'
+# ]^M
 _base_ = [
-    '../../../../configs/_base_/datasets/nus-3d.py',
-    '../../../../configs/_base_/default_runtime.py'
+    '../../../../plugin/_base_/datasets/nus-3d.py',
+    '../../../../plugin/_base_/default_runtime.py'
 ]
 plugin=True
 plugin_dir='plugin/futr3d/'
@@ -288,7 +292,7 @@ data = dict(
         # dataset=dict(
             type=dataset_type,
             data_root=data_root,
-            ann_file=data_root + 'radar_nuscenes_5sweeps_infos_train_radar.pkl',
+            ann_file=data_root + 'radar_nuscenes_5sweeps_infos_train.pkl',
             pipeline=train_pipeline,
             classes=class_names,
             modality=input_modality,
@@ -301,11 +305,11 @@ data = dict(
     val=dict(
             type=dataset_type,
             pipeline=test_pipeline, classes=class_names, modality=input_modality,
-            ann_file=data_root + 'radar_nuscenes_5sweeps_infos_val_radar.pkl',),
+            ann_file=data_root + 'radar_nuscenes_5sweeps_infos_val.pkl',),^M
     test=dict(
             type=dataset_type,
             pipeline=test_pipeline, classes=class_names, modality=input_modality,
-            ann_file=data_root + 'radar_nuscenes_5sweeps_infos_val_radar.pkl',),)
+            ann_file=data_root + 'radar_nuscenes_5sweeps_infos_val.pkl',),)^M
 
 optimizer = dict(
     type='AdamW',
diff --git a/plugin/futr3d/models/backbones/radar_encoder.py b/plugin/futr3d/models/backbones/radar_encoder.py
index 9866afb..fc10d5b 100644
--- a/plugin/futr3d/models/backbones/radar_encoder.py
+++ b/plugin/futr3d/models/backbones/radar_encoder.py
@@ -78,6 +78,8 @@ class RadarPointEncoder(nn.Module):
         ret: 
             out: [B, N, C+1], last channel as 0-1 mask
         '''
+        if isinstance(points,list):^M
+            points = points[0]^M
         masks = points[:, :, [-1]]
         x = points[:, :, :-1]
         xy = points[:, :, :2]
@@ -90,4 +92,4 @@ class RadarPointEncoder(nn.Module):
         out = torch.cat((x, masks), dim=-1)
 
         out = torch.cat((xy, out), dim=-1)
-        return 
\ No newline at end of file
+        return out
\ No newline at end of file
diff --git a/plugin/futr3d/models/utils/attention.py b/plugin/futr3d/models/utils/attention.py
index a945a90..af59e08 100644
--- a/plugin/futr3d/models/utils/attention.py
+++ b/plugin/futr3d/models/utils/attention.py
@@ -267,6 +267,7 @@ class FUTR3DCrossAtten(BaseModule):
             pts_output = self.pts_output_proj(pts_output)
         
         if self.use_Radar:
+            radar_feats = rad_feats
             radar_feats, radar_mask = radar_feats[:, :, :-1], radar_feats[:, :, -1]
             radar_xy = radar_feats[:, :, :2]
             ref_xy = reference_points[:, :, :2]
diff --git a/tools/data_converter/nusc_radar.py b/tools/data_converter/nusc_radar.py
index edb18f0..709c119 100644
--- a/tools/data_converter/nusc_radar.py
+++ b/tools/data_converter/nusc_radar.py
@@ -640,4 +640,4 @@ def generate_record(ann_rec: dict, x1: float, y1: float, x2: float, y2: float,
 if __name__ == '__main__':
     # create_nuscenes_infos('data/nuscenes/', 'track_radar')
 
-    create_nuscenes_infos('data/nuscenes/', 'radar_nuscenes_5sweeps', version='v1.0-test')
+    create_nuscenes_infos('data/nuscenes/', 'radar_nuscenes_5sweeps', version='v1.0-trainval') # 'v1.0-test')
diff --git a/tools/dist_test.sh b/tools/dist_test.sh
index 3c74ec6..abb79ef 100644
--- a/tools/dist_test.sh
+++ b/tools/dist_test.sh
@@ -7,4 +7,4 @@ PORT=${PORT:-29500}
 
 PYTHONPATH="$(dirname $0)/..":$PYTHONPATH \
 python -m torch.distributed.launch --nproc_per_node=$GPUS --master_port=$PORT \
-    $(dirname "$0")/test.py $CONFIG $CHECKPOINT --launcher pytorch ${@:4}
+    $(dirname "$0")/test.py $CONFIG $CHECKPOINT --launcher pytorch ${@:4} --eval mAP

```
# run command
```
bash tools/dist_test.sh /data/futr3d/plugin/futr3d/configs/cam_radar/res101_radar.py ./cam_radar.pth 1
```
 Hi, @xyaochen ,
I changed the codebase like above and run the eval code,the info is shown as below,the mAP is 0,is anywhere fault?
```
[                                                  ] 1/6019, 0.3 task/s, elapsed: 4s, ETA: 21446s/data/futr3d/plugin/futr3d/core/bbox/coders/nms_free_coder.py:71: UserWarning: To copy construct from a ten
sor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  self.post_center_range, device=scores.device)
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 6019/6019, 2.1 task/s, elapsed: 2882s, ETA:     0s
Formating bboxes of pts_bbox
Start to convert detection format...
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 6019/6019, 15.7 task/s, elapsed: 384s, ETA:     0s
Results writes to /tmp/tmpt4ggv_c0/results/pts_bbox/results_nusc.json
Evaluating bboxes of pts_bbox
mAP: 0.0000
mATE: 1.0277
mASE: 0.9499
mAOE: 1.0682
mAVE: 0.9894
mAAE: 0.9631
NDS: 0.0098
Eval time: 236.7s

Per-class results:
Object Class    AP      ATE     ASE     AOE     AVE     AAE
car     0.000   1.000   1.000   1.000   1.000   1.000
truck   0.000   1.000   1.000   1.000   1.000   1.000
bus     0.000   1.000   1.000   1.000   1.000   1.000
trailer 0.000   1.000   1.000   1.000   1.000   1.000
construction_vehicle    0.000   1.000   1.000   1.000   1.000   1.000
pedestrian      0.000   1.277   0.499   1.614   0.915   0.705
motorcycle      0.000   1.000   1.000   1.000   1.000   1.000
bicycle 0.000   1.000   1.000   1.000   1.000   1.000
traffic_cone    0.000   1.000   1.000   nan     nan     nan
barrier 0.000   1.000   1.000   1.000   nan     nan
{'pts_bbox_NuScenes/car_AP_dist_0.5': 0.0, 'pts_bbox_NuScenes/car_AP_dist_1.0': 0.0, 'pts_bbox_NuScenes/car_AP_dist_2.0': 0.0, 'pts_bbox_NuScenes/car_AP_dist_4.0': 0.0, 'pts_bbox_NuScenes/car_trans_err':
1.0, 'pts_bbox_NuScenes/car_scale_err': 1.0, 'pts_bbox_NuScenes/car_orient_err': 1.0, 'pts_bbox_NuScenes/car_vel_err': 1.0, 'pts_bbox_NuScenes/car_attr_err': 1.0, 'pts_bbox_NuScenes/truck_AP_dist_0.5': 0.
0, 'pts_bbox_NuScenes/truck_AP_dist_1.0': 0.0, 'pts_bbox_NuScenes/truck_AP_dist_2.0': 0.0, 'pts_bbox_NuScenes/truck_AP_dist_4.0': 0.0, 'pts_bbox_NuScenes/truck_trans_err': 1.0, 'pts_bbox_NuScenes/truck_sc
ale_err': 1.0, 'pts_bbox_NuScenes/truck_orient_err': 1.0, 'pts_bbox_NuScenes/truck_vel_err': 1.0, 'pts_bbox_NuScenes/truck_attr_err': 1.0, 'pts_bbox_NuScenes/construction_vehicle_AP_dist_0.5': 0.0, 'pts_b
box_NuScenes/construction_vehicle_AP_dist_1.0': 0.0, 'pts_bbox_NuScenes/construction_vehicle_AP_dist_2.0': 0.0, 'pts_bbox_NuScenes/construction_vehicle_AP_dist_4.0': 0.0, 'pts_bbox_NuScenes/construction_v
ehicle_trans_err': 1.0, 'pts_bbox_NuScenes/construction_vehicle_scale_err': 1.0, 'pts_bbox_NuScenes/construction_vehicle_orient_err': 1.0, 'pts_bbox_NuScenes/construction_vehicle_vel_err': 1.0, 'pts_bbox_
NuScenes/construction_vehicle_attr_err': 1.0, 'pts_bbox_NuScenes/bus_AP_dist_0.5': 0.0, 'pts_bbox_NuScenes/bus_AP_dist_1.0': 0.0, 'pts_bbox_NuScenes/bus_AP_dist_2.0': 0.0, 'pts_bbox_NuScenes/bus_AP_dist_4
.0': 0.0, 'pts_bbox_NuScenes/bus_trans_err': 1.0, 'pts_bbox_NuScenes/bus_scale_err': 1.0, 'pts_bbox_NuScenes/bus_orient_err': 1.0, 'pts_bbox_NuScenes/bus_vel_err': 1.0, 'pts_bbox_NuScenes/bus_attr_err': 1
.0, 'pts_bbox_NuScenes/trailer_AP_dist_0.5': 0.0, 'pts_bbox_NuScenes/trailer_AP_dist_1.0': 0.0, 'pts_bbox_NuScenes/trailer_AP_dist_2.0': 0.0, 'pts_bbox_NuScenes/trailer_AP_dist_4.0': 0.0, 'pts_bbox_NuScen
es/trailer_trans_err': 1.0, 'pts_bbox_NuScenes/trailer_scale_err': 1.0, 'pts_bbox_NuScenes/trailer_orient_err': 1.0, 'pts_bbox_NuScenes/trailer_vel_err': 1.0, 'pts_bbox_NuScenes/trailer_attr_err': 1.0, 'p
ts_bbox_NuScenes/barrier_AP_dist_0.5': 0.0, 'pts_bbox_NuScenes/barrier_AP_dist_1.0': 0.0, 'pts_bbox_NuScenes/barrier_AP_dist_2.0': 0.0, 'pts_bbox_NuScenes/barrier_AP_dist_4.0': 0.0, 'pts_bbox_NuScenes/bar
rier_trans_err': 1.0, 'pts_bbox_NuScenes/barrier_scale_err': 1.0, 'pts_bbox_NuScenes/barrier_orient_err': 1.0, 'pts_bbox_NuScenes/barrier_vel_err': nan, 'pts_bbox_NuScenes/barrier_attr_err': nan, 'pts_bbo
x_NuScenes/motorcycle_AP_dist_0.5': 0.0, 'pts_bbox_NuScenes/motorcycle_AP_dist_1.0': 0.0, 'pts_bbox_NuScenes/motorcycle_AP_dist_2.0': 0.0, 'pts_bbox_NuScenes/motorcycle_AP_dist_4.0': 0.0, 'pts_bbox_NuScen
es/motorcycle_trans_err': 1.0, 'pts_bbox_NuScenes/motorcycle_scale_err': 1.0, 'pts_bbox_NuScenes/motorcycle_orient_err': 1.0, 'pts_bbox_NuScenes/motorcycle_vel_err': 1.0, 'pts_bbox_NuScenes/motorcycle_att
r_err': 1.0, 'pts_bbox_NuScenes/bicycle_AP_dist_0.5': 0.0, 'pts_bbox_NuScenes/bicycle_AP_dist_1.0': 0.0, 'pts_bbox_NuScenes/bicycle_AP_dist_2.0': 0.0, 'pts_bbox_NuScenes/bicycle_AP_dist_4.0': 0.0, 'pts_bb
ox_NuScenes/bicycle_trans_err': 1.0, 'pts_bbox_NuScenes/bicycle_scale_err': 1.0, 'pts_bbox_NuScenes/bicycle_orient_err': 1.0, 'pts_bbox_NuScenes/bicycle_vel_err': 1.0, 'pts_bbox_NuScenes/bicycle_attr_err'
: 1.0, 'pts_bbox_NuScenes/pedestrian_AP_dist_0.5': 0.0, 'pts_bbox_NuScenes/pedestrian_AP_dist_1.0': 0.0, 'pts_bbox_NuScenes/pedestrian_AP_dist_2.0': 0.0, 'pts_bbox_NuScenes/pedestrian_AP_dist_4.0': 0.0, '
pts_bbox_NuScenes/pedestrian_trans_err': 1.2771, 'pts_bbox_NuScenes/pedestrian_scale_err': 0.4987, 'pts_bbox_NuScenes/pedestrian_orient_err': 1.6141, 'pts_bbox_NuScenes/pedestrian_vel_err': 0.9152, 'pts_b
box_NuScenes/pedestrian_attr_err': 0.7047, 'pts_bbox_NuScenes/traffic_cone_AP_dist_0.5': 0.0, 'pts_bbox_NuScenes/traffic_cone_AP_dist_1.0': 0.0, 'pts_bbox_NuScenes/traffic_cone_AP_dist_2.0': 0.0, 'pts_bbo
x_NuScenes/traffic_cone_AP_dist_4.0': 0.0, 'pts_bbox_NuScenes/traffic_cone_trans_err': 1.0, 'pts_bbox_NuScenes/traffic_cone_scale_err': 1.0, 'pts_bbox_NuScenes/traffic_cone_orient_err': nan, 'pts_bbox_NuS
cenes/traffic_cone_vel_err': nan, 'pts_bbox_NuScenes/traffic_cone_attr_err': nan, 'pts_bbox_NuScenes/NDS': 0.009765414502821469, 'pts_bbox_NuScenes/mAP': 0.0}
```

**Comment by Song-Jingyu** [comment]
Want to follow up here as the code looks a little weird here to me,

```
        out = x * masks

        out = torch.cat((x, masks), dim=-1)

        out = torch.cat((xy, out), dim=-1)
        return 
```

What is the motivation of out=x*masks here since the out got overwrite later on. Thanks!
